pkg/expapi/latest/latest.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package latest
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/meta"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/registered"
+	_ "github.com/GoogleCloudPlatform/kubernetes/pkg/expapi"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/expapi/v1"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+)
+
+var (
+	Version  string
+	Versions []string
+
+	accessor   = meta.NewAccessor()
+	Codec      runtime.Codec
+	SelfLinker = runtime.SelfLinker(accessor)
+	RESTMapper meta.RESTMapper
+)
+
+func init() {
+	Version = registered.RegisteredVersions[0]
+	Codec = runtime.CodecFor(api.Scheme, Version)
+	// Put the registered versions in Versions in reverse order.
+	for i := len(registered.RegisteredVersions) - 1; i >= 0; i-- {
+		Versions = append(Versions, registered.RegisteredVersions[i])
+	}
+
+	mapper := meta.NewDefaultRESTMapper(
+		Versions,
+		func(version string) (*meta.VersionInterfaces, bool) {
+			interfaces, err := InterfacesFor(version)
+			if err != nil {
+				return nil, false
+			}
+			return interfaces, true
+		},
+	)
+
+	// the list of kinds that are scoped at the root of the api hierarchy
+	// if a kind is not enumerated here, it is assumed to have a namespace scope
+	rootScoped := map[string]bool{}
+
+	ignoredKinds := util.NewStringSet()
+
+	for _, version := range Versions {
+		for kind, oType := range api.Scheme.KnownTypes(version) {
+			if !strings.Contains(oType.PkgPath(), "kubernetes/pkg/expapi") || ignoredKinds.Has(kind) {
+				continue
+			}
+			scope := meta.RESTScopeNamespace
+			if rootScoped[kind] {
+				scope = meta.RESTScopeRoot
+			}
+			mapper.Add(scope, kind, version, false)
+		}
+	}
+	RESTMapper = mapper
+	api.RegisterRESTMapper(RESTMapper)
+}
+
+// InterfacesFor returns the default Codec and ResourceVersioner for a given version
+// string, or an error if the version is not known.
+func InterfacesFor(version string) (*meta.VersionInterfaces, error) {
+	switch version {
+	case "v1":
```

Comment:
We have a global registry of versions (pkg/api/registry) and dealing with separate versions for experimental and api is icky enough that I would rather it be changed after we have multiple groups.